src/rdkafka_telemetry_encode.c

```diff
@@ -0,0 +1,834 @@
+        rd_kafka_telemetry_metric_value_t max_rtt;
+
+        max_rtt.int_value = RD_CEIL_INTEGER_DIVISION(
+            rkb_selected->rkb_telemetry.rd_avg_rollover.rkb_avg_rtt.ra_v.maxv,
+            THREE_ORDERS_MAGNITUDE);
+        return max_rtt;
+}
+
+static rd_kafka_telemetry_metric_value_t
+calculate_throttle_avg(rd_kafka_t *rk,
+                       rd_kafka_broker_t *rkb_selected,
+                       rd_ts_t now_ns) {
+        rd_kafka_telemetry_metric_value_t avg_throttle;
+        rd_kafka_broker_t *rkb;
+        double avg = 0;
+        int count  = 0;
+
+        TAILQ_FOREACH(rkb, &rk->rk_brokers, rkb_link) {
+                rd_avg_t *rkb_avg_throttle_rollover =
+                    &rkb->rkb_telemetry.rd_avg_rollover.rkb_avg_throttle;
+                if (rkb_avg_throttle_rollover->ra_v.cnt) {
+                        avg = (avg * count +
+                               rkb_avg_throttle_rollover->ra_v.sum) /
+                              (double)(count +
+                                       rkb_avg_throttle_rollover->ra_v.cnt);
+                        count += rkb_avg_throttle_rollover->ra_v.cnt;
+                }
+        }
+        avg_throttle.double_value = avg;
+        return avg_throttle;
+}
+
+
+static rd_kafka_telemetry_metric_value_t
+calculate_throttle_max(rd_kafka_t *rk,
+                       rd_kafka_broker_t *rkb_selected,
+                       rd_ts_t now_ns) {
+        rd_kafka_telemetry_metric_value_t max_throttle;
+        rd_kafka_broker_t *rkb;
+
+        max_throttle.int_value = 0;
+        TAILQ_FOREACH(rkb, &rk->rk_brokers, rkb_link) {
+                max_throttle.int_value = RD_MAX(
+                    max_throttle.int_value, rkb->rkb_telemetry.rd_avg_rollover
+                                                .rkb_avg_throttle.ra_v.maxv);
+        }
+        return max_throttle;
+}
+
+static rd_kafka_telemetry_metric_value_t
+calculate_queue_time_avg(rd_kafka_t *rk,
+                         rd_kafka_broker_t *rkb_selected,
+                         rd_ts_t now_ns) {
+        rd_kafka_telemetry_metric_value_t avg_queue_time;
+        rd_kafka_broker_t *rkb;
+        double avg = 0;
+        int count  = 0;
+
+        TAILQ_FOREACH(rkb, &rk->rk_brokers, rkb_link) {
+                rd_avg_t *rkb_avg_outbuf_latency_rollover =
+                    &rkb->rkb_telemetry.rd_avg_rollover.rkb_avg_outbuf_latency;
+                if (rkb_avg_outbuf_latency_rollover->ra_v.cnt) {
+                        avg =
+                            (avg * count +
+                             rkb_avg_outbuf_latency_rollover->ra_v.sum) /
+                            (double)(count +
+                                     rkb_avg_outbuf_latency_rollover->ra_v.cnt);
+                        count += rkb_avg_outbuf_latency_rollover->ra_v.cnt;
+                }
+        }
+
+        avg_queue_time.double_value =
+            RD_CEIL_INTEGER_DIVISION(avg, THREE_ORDERS_MAGNITUDE);
```

Comment:
As a floating point division there's no need to take the ceil.
```suggestion
        avg_queue_time.double_value = avg / THREE_ORDERS_MAGNITUDE;
```

Reply:
Updated